### ISSUE TITLE
Double back button clicks needed for navigating back in central dashboard

### ISSUE THREAD

**Issue by Bobgy** [kubeflow/kubeflow#4446]
/kind bug
/area centraldashboard

**What steps did you take and what happened:**
In KFP run detail page, I need to click back button twice to go back to run list page. It's similar for other pages including Katib and notebook servers.
https://screencast.googleplex.com/cast/NDYyNzM2NjM0MzA4MTk4NHxhZjI5MDBlMC05Zg

**What did you expect to happen:**
Single click on back button should be enough.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- Kubeflow version: (version number can be found at the bottom left corner of the Kubeflow dashboard): build version 0.7.0
- kfctl version: (use `kfctl version`): kfctl v0.7.0-rc.5-27-g7f64d8b0
- Kubernetes platform: (e.g. `minikube`) gcp-iap
- Kubernetes version: (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.0", GitCommit:"2bd9643cee5b3b3a5ecbd3af49d09018f0773c77", GitTreeState:"clean", BuildDate:"2019-09-18T14:36:53Z", GoVersion:"go1.12.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.10-gke.0", GitCommit:"569511c9540f78a94cc6a41d895c382d0946c11a", GitTreeState:"clean", BuildDate:"2019-08-21T23:28:44Z", GoVersion:"go1.11.13b4", Compiler:"gc", Platform:"linux/amd64"}
```
- OS (e.g. from `/etc/os-release`): both for linux and mac

**Comment by Bobgy** [comment]
@prodonjs @avdaredevil do you have any context?

**Comment by prodonjs** [comment]
@Bobgy - We did have this regression but addressed it in #4326 which should be included in the rc-5 candidate you indicated. It looks like the issue might be somewhat different though than the core one that we wanted to address. Both Katib and KFP use hash based routing and the back button you reference in KFP simply uses the browsers's history functionality to go back a page.

When using KFP or any other KF app inside the context of the centraldashboard's iframe, we added code that listens for any clicks in the iframe and updates the parent page's URL when the location changes. So if I click on the Pipelines link in the left nav menu, the top-level path becomes `/_/pipeline/`. Then, if I click a Pipline definition, it becomes `/_/pipeline/#/pipelines/details/<pipeline_id>`. Then, clicking the button to go back to the list page calls history.goBack(), but my theory is that due to the iframe context, the back operation is affecting the overall browser's URL and therefore dropping the # section of the URL and changing it back to `/_/pipeline/`. To the centraldashboard, this is not a navigation event since it doesn't use hash-based routing. I believe what happens is that the dashboard tries to reset the iframe's source to the new URL that no longer contains the #, but that doesn't trigger a reload of the KFP page so it's as if the navigation event didn't happen.

A quick test of this theory would be to see if changing the functionality of the back button in KFP to use a standard anchor tag to the list page fixes the issue. In general, it's been a bit of a challenge to handle some of the nuances of iframing without explicitly establishing a channel for apps and the centraldashboard to use. Hopefully this is something we can address without major changes.

**Comment by prodonjs** [comment]
I can confirm that this is a regression based on the changes made in #4230 to account for navigation changes in the iframed-child pages to the URL of the parent as the v0.6.2 version of the dashboard does not exhibit the same behavior.

**Comment by Bobgy** [comment]
Just two cents from my understanding, haven't tried it out.
IFrame history changes are already records in window history, ref: http://www.webdeveasy.com/back-button-behavior-on-a-page-with-an-iframe/
So when we sync window location to iframe, we shouldn't use `pushState` which creates a second entry in https://github.com/kubeflow/kubeflow/pull/4230/files#diff-5184304ce3ffddcc77cdade1f1140072R250.

Instead, we should use `replaceState`, ref: https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState

**Comment by prodonjs** [comment]
@Bobgy - That's a great suggestion. I think that should resolve the issue so I'll test it tomorrow morning and get a PR ready if it works as expected.

**Comment by prodonjs** [comment]
Fix is ready in PR #4454. If you want to verify it, you can try image gcr.io/prodonjs-kubeflow-dev/centraldashboard@sha256:702fe888e94d4596b353ea9002aa0b8ac6818d1d5573bf9c37320c2d05f5b0cf
